qi-doc/scribblings/macros.scrbl

```diff
@@ -0,0 +1,90 @@
+#lang scribble/doc
+@require[scribble/manual
+         scribble-abbrevs/manual
+         scribble/example
+         racket/sandbox
+         @for-label[qi
+                    racket
+                    syntax/parse
+                    syntax/parse/define
+                    (only-in relation
+                             ->number
+                             ->string
+                             sum)]]
+
+@(define eval-for-docs
+  (parameterize ([sandbox-output 'string]
+                 [sandbox-error-output 'string]
+                 [sandbox-memory-limit #f])
+    (make-evaluator 'racket/base
+                    '(require qi
+                              (only-in racket/list range first rest)
+                              (for-syntax syntax/parse racket/base)
+                              racket/string
+                              relation)
+                    '(define (sqr x)
+                       (* x x)))))
+
+@title{Qi Macros}
+
+Qi may be extended in much the same way as Racket -- using @tech/reference{macros}. Qi macros are "first class," meaning that they are indistinguishable from built-in Qi forms during the macro expansion phase, just as user-defined Racket macros as indistinguishable from macros that are part of the Racket language. This allows us to have the same syntactic freedom with Qi as we are used to with Racket.
```

Comment:
I'm not sure whether "first class" is an appropriate term. "First class" is commonly used in the PL literature to mean that a kind of thing may be used as a value in the language. And indeed, some languages have actual "first-class" macros, such that something like the following program would be legal and would terminate:

```
(define-syntax-rule (delay e) (lambda () e))
(define (f g) (g (let loop () (loop))))
(f delay)
```

Such macros (like FEXPRs which are related but not the same) require that expansion be part of evaluation, as opposed to a separate step ahead of time.

I'm not sure of a better term however. Is it important to convey that Qi macros are "first-class" in the sense you mean?

Reply:
True, I did think that it might be interpreted in this sense but I wasn't sure if the term was actually used in this way. What do you think about using the term "first class macro extensibility" but not "first class macros"?

And nice, TIL about FEXPRs and the possibility of simultaneous expansion + evaluation!